src/azure/azureLogin/azureAuth.ts

```diff
@@ -0,0 +1,150 @@
+            title: "Waiting for sign-in",
+            cancellable: true,
+        };
+
+        await vscode.window.withProgress(options, (_, token) => {
+            let listener: vscode.Disposable | undefined;
+            token.onCancellationRequested(listener?.dispose());
+            return new Promise((resolve) => {
+                listener = sessionProvider.signInStatusChangeEvent((status) => {
+                    if (status === SignInStatus.SignedIn) {
+                        listener?.dispose();
+                        resolve(undefined);
+                    }
+                });
+            });
+        });
+    }
+
+    export function getConfiguredAzureEnv(): Environment {
+        // See:
+        // https://github.com/microsoft/vscode/blob/eac16e9b63a11885b538db3e0b533a02a2fb8143/extensions/microsoft-authentication/package.json#L40-L99
+        const section = "microsoft-sovereign-cloud";
+        const settingName = "environment";
+        const authProviderConfig = vscode.workspace.getConfiguration(section);
+        const environmentSettingValue = authProviderConfig.get<string | undefined>(settingName);
+
+        if (environmentSettingValue === "ChinaCloud") {
+            return Environment.ChinaCloud;
+        } else if (environmentSettingValue === "USGovernment") {
+            return Environment.USGovernment;
+        } else if (environmentSettingValue === "custom") {
+            const customCloud = authProviderConfig.get<EnvironmentParameters | undefined>("customEnvironment");
+            if (customCloud) {
+                return new Environment(customCloud);
+            }
+
+            throw new Error(
+                `The custom cloud choice is not configured. Please configure the setting ${section}.${settingName}.`,
+            );
+        }
+
+        return Environment.get(Environment.AzureCloud.name);
+    }
+
+    export async function listAll<T>(iterator: PagedAsyncIterableIterator<T>): Promise<Utils.Errorable<T[]>> {
+        const result: T[] = [];
+        try {
+            for await (const page of iterator.byPage()) {
+                result.push(...page);
+            }
+            return { succeeded: true, result };
+        } catch (e) {
+            return { succeeded: false, error: `Failed to list resources: ${Utils.getErrorMessage(e)}` };
+        }
+    }
+
+    export function getScopes(tenantId: string | null, options: GetAuthSessionOptions): string[] {
+        const defaultScopes = options.scopes || [AzureAuth.getDefaultScope(AzureAuth.getConfiguredAzureEnv().resourceManagerEndpointUrl)];
+        const tenantScopes = tenantId ? [`VSCODE_TENANT:${tenantId}`] : [];
+        const clientIdScopes = options.applicationClientId ? [`VSCODE_CLIENT_ID:${options.applicationClientId}`] : [];
+        return [...defaultScopes, ...tenantScopes, ...clientIdScopes];
+    }
+
+    type AuthProviderId = "microsoft" | "microsoft-sovereign-cloud";
```

Comment:
what is microsoft-sovereign-cloud?

Reply:
azure subscriptions for special customers, such as azure china, azure government and so on.
https://blogs.microsoft.com/blog/2023/12/14/microsoft-cloud-for-sovereignty-now-generally-available-opening-new-pathways-for-government-innovation/